beater/gpfs.go

```diff
@@ -89,56 +126,142 @@ func parseCertainInt(s string) int64 {
+
+		// there may be multiple HEADER lines so we need to gather them here (which is ugly, granted)
+		// we then also already have the line identifier, so no need to get it from the HEADER parsing
+		if updateHeaderMap(headerFieldLocation, identifierFieldLocation, headerMap, fields) {
+			continue // we updated the map, so we can skip the remainder for this header line
 		}
-		qs = append(qs, qi)
+		identifier := fields[identifierFieldLocation]
+		fieldMap := headerMap[identifier]
+		info := fn(fields, fieldMap)
+		result = append(result, info)
 	}
-	return qs, nil
+
+	return result, nil
 }
 
-// parseMmRepQuotaHeader builds a map of the field names and the corresponding index
-func parseMmRepQuotaHeader(header string) map[string]int {
+// parseMmLsFs returns the different devices in the GPFS cluster
+func parseMmLsFs(output string) ([]string, error) {
+	var prefixFieldlocation = 0
+	var identifierFieldLocation = 1
+	var headerFieldLocation = 2
+
+	ds, _ := parseGpfsYOutput(prefixFieldlocation, identifierFieldLocation, headerFieldLocation, "mmlsfs", output, parseMmLsFsCallback)
+
+	// avoid duplicates
+	var dsm map[string]bool
+	var devices = make([]string, 0, len(ds))
+	for _, device := range ds {
+		d := device.(string) // explicit type conversion
+		_, ok := dsm[d]
+		if !ok {
+			dsm[d] = true
+			devices = append(devices, d)
+		}
+	}
+
+	return devices, nil
+}
+
+// ParseMmRepQuota converts the lines into the desired information
+func parseMmRepQuota(output string) ([]QuotaInfo, error) {
+
+	var prefixFieldlocation = 0
+	var identifierFieldLocation = 1
+	var headerFieldLocation = 2
 
-	var m = make(map[string]int)
+	qs, _ := parseGpfsYOutput(prefixFieldlocation, identifierFieldLocation, headerFieldLocation, "mmrepquota", output, parseMmRepQuotaCallback)
+
+	var quotaInfos = make([]QuotaInfo, 0, len(qs))
+	for _, q := range qs {
+		quotaInfos = append(quotaInfos, q.(QuotaInfo))
+	}
+
+	return quotaInfos, nil
+}
+
+func parseMmRepQuotaCallback(fields []string, fieldMap map[string]int) interface{} {
+	qi := QuotaInfo{
+		filesystem: fields[fieldMap["filesystemName"]],
+		fileset:    fields[fieldMap["filesetname"]],
+		kind:       fields[fieldMap["quotaType"]],
+		entity:     fields[fieldMap["name"]],
+		blockUsage: parseCertainInt(fields[fieldMap["blockUsage"]]),
+		blockSoft:  parseCertainInt(fields[fieldMap["blockQuota"]]),
+		blockHard:  parseCertainInt(fields[fieldMap["blockLimit"]]),
+		blockDoubt: parseCertainInt(fields[fieldMap["blockInDoubt"]]),
+		blockGrace: fields[fieldMap["blockGrace"]],
+		filesUsage: parseCertainInt(fields[fieldMap["filesUsage"]]),
+		filesSoft:  parseCertainInt(fields[fieldMap["filesQuota"]]),
+		filesHard:  parseCertainInt(fields[fieldMap["filesLimit"]]),
+		filesDoubt: parseCertainInt(fields[fieldMap["filesInDoubt"]]),
+		filesGrace: fields[fieldMap["filesGrace"]],
+	}
+	if qi.kind == "FILESET" {
+		qi.fileset = qi.entity // filesets have no name, and we need to have a link between FILESET and USR quota
```

Comment:
`name` -> `filesetname attribute`
this could be a bug.
@kwaegema feel like a new ticket: in the output of `mmreqpuota -Y`, the `filesetname` is not set for lines of type `FILESET`.